discovery-provider/src/tasks/index_user_bank.py

```diff
@@ -0,0 +1,309 @@
+            if not transactions_array:
+                intersection_found = True
+                logger.info(
+                    f"index_user_bank.py | No transactions found before {last_tx_signature}"
+                )
+            else:
+                # Current batch of transactions
+                transaction_signature_batch = []
+                for tx_info in transactions_array:
+                    tx_sig = tx_info["signature"]
+                    tx_slot = tx_info["slot"]
+                    logger.info(
+                        f"index_user_bank.py | Processing tx={tx_sig} | slot={tx_slot}"
+                    )
+                    if tx_info["slot"] > latest_processed_slot:
+                        transaction_signature_batch.append(tx_sig)
+                    elif (
+                        tx_info["slot"] <= latest_processed_slot
+                        and tx_info["slot"] > MIN_SLOT
+                    ):
+                        # Check the tx signature for any txs in the latest batch,
+                        # and if not present in DB, add to processing
+                        logger.info(
+                            f"index_user_bank.py | Latest slot re-traversal\
+                            slot={tx_slot}, sig={tx_sig},\
+                            latest_processed_slot(db)={latest_processed_slot}"
+                        )
+                        exists = get_tx_in_db(session, tx_sig)
+                        if exists:
+                            intersection_found = True
+                            break
+                        # Ensure this transaction is still processed
+                        transaction_signature_batch.append(tx_sig)
+
+                # Restart processing at the end of this transaction signature batch
+                last_tx = transactions_array[-1]
+                last_tx_signature = last_tx["signature"]
+
+                # Append batch of processed signatures
+                if transaction_signature_batch:
+                    transaction_signatures.append(transaction_signature_batch)
+
+                # Ensure processing does not grow unbounded
+                if len(transaction_signatures) > TX_SIGNATURES_MAX_BATCHES:
+                    prev_len = len(transaction_signatures)
+                    # Only take the oldest transaction from the transaction_signatures array
+                    # transaction_signatures is sorted from newest to oldest
+                    transaction_signatures = transaction_signatures[
+                        -TX_SIGNATURES_RESIZE_LENGTH:
+                    ]
+                    logger.info(
+                        f"index_user_bank.py | sliced tx_sigs from {prev_len} to {len(transaction_signatures)} entries"
+                    )
+
+    # Reverse batches aggregated so oldest transactions are processed first
+    transaction_signatures.reverse()
+
+    num_txs_processed = 0
+    for tx_sig_batch in transaction_signatures:
```

Comment:
I don't think we have to do this now, but if we need a perf gain we could restructure this a bit - right now we spin up 5 new executors, and acquire 5 sessions for each batch that we process. What we could do instead is slice the batch in 5 parts to begin with, and then give each executor a session and 1/5 of the array to chew thru. that we way don't incur the cost of dealing with threads and sessions for each signature batch

Reply:
this is a fantastic comment, let's add it in follow-up.